### PR TITLE
feat: Advanced Contact Filters

### LINKED ISSUE #3190: Contacts view Filter.
Use the component in #3188 to generate the filter payload for contacts and send it in the contacts API to return custom segments of contacts.

### PR DESCRIPTION
**Pull request by fayazara**
# Pull Request Template

## Description

Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes #3190

## Type of change

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

## How Has This Been Tested?

Test specs for actions, mutations and getters. Tested on staging env.

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

https://user-images.githubusercontent.com/15716057/143827837-72005167-66f9-40d6-8ef2-729cc1005859.mp4

### CODE REVIEW

File: app/javascript/dashboard/components/widgets/conversation/ConversationAdvancedFilter.vue [hunk 1]
```diff
@@ -48,9 +48,9 @@
 <script>
 import alertMixin from 'shared/mixins/alertMixin';
 import { required, requiredIf } from 'vuelidate/lib/validators';
-import FilterInputBox from './components/FilterInput.vue';
+import FilterInputBox from '../FilterInput.vue';
 import languages from './advancedFilterItems/languages';
-import countries from './advancedFilterItems/countries';
+import countries from '../../../../shared/constants/countries';
```
Comment (nithindavid) [inline app/javascript/dashboard/components/widgets/conversation/ConversationAdvancedFilter.vue:53]: We could use absolute path here like `app/javascript/dashboard/components/MyLib`

File: app/javascript/dashboard/components/widgets/conversation/ConversationAdvancedFilter.vue [hunk 2]
```diff
@@ -93,19 +93,18 @@ export default {
...
-          name: type.attributeName,
-          attributeI18nKey: type.attributeI18nKey,
+          name: this.$t(`FILTER.ATTRIBUTES.${type.attributeI18nKey}`),
         };
       });
     },
-    getAppliedFilters() {
-      return this.$store.getters.getAppliedFilters;
+    getAppliedConversationFilters() {
+      return this.$store.getters.getAppliedConversationFilters;
```
Comment (nithindavid) [inline app/javascript/dashboard/components/widgets/conversation/ConversationAdvancedFilter.vue:101]: We could use mapGetters ?

**Review by nithindavid — CHANGES_REQUESTED** [review]
Have some minor comments

File: app/javascript/dashboard/components/widgets/conversation/ConversationAdvancedFilter.vue (hunk 2 above)
Comment (fayazara) [reply app/javascript/dashboard/components/widgets/conversation/ConversationAdvancedFilter.vue]: Fixed

File: app/javascript/dashboard/components/widgets/conversation/ConversationAdvancedFilter.vue (hunk 1 above)
Comment (fayazara) [reply app/javascript/dashboard/components/widgets/conversation/ConversationAdvancedFilter.vue]: Fixed

File: app/javascript/dashboard/routes/dashboard/contacts/components/ContactsAdvancedFilters.vue [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+      const type = this.filterTypes.find(filter => filter.attributeKey === key);
+      return type.filterOperators;
+    },
+    // eslint-disable-next-line consistent-return
+    getDropdownValues(type) {
+      switch (type) {
+        case 'country_code':
+          return countries;
+        default:
+          break;
```
Comment (nithindavid) [inline app/javascript/dashboard/routes/dashboard/contacts/components/ContactsAdvancedFilters.vue:139]: Or default value to get rid of the eslint warning
```suggestion
          return undefined;
```

File: app/javascript/dashboard/routes/dashboard/contacts/components/ContactsAdvancedFilters.vue [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+      }
+    },
+    appendNewFilter() {
+      this.appliedFilters.push({
+        attribute_key: 'name',
```
Comment (nithindavid) [inline app/javascript/dashboard/routes/dashboard/contacts/components/ContactsAdvancedFilters.vue:144]: Usually we avoid having snake keys on the JS, If API needs it we change it either on the axios API functions.  So we could avoid `query_operator` in the code

File: app/javascript/dashboard/routes/dashboard/contacts/components/Header.vue [hunk 5]
```diff
@@ -155,4 +181,17 @@ export default {
     visibility: visible;
   }
 }
+.filters__button-wrap {
+  position: relative;
+
+  .filters__applied-indicator {
```
Comment (nithindavid) [inline app/javascript/dashboard/routes/dashboard/contacts/components/Header.vue:187]: Use variables

File: app/javascript/dashboard/routes/dashboard/contacts/components/ContactsAdvancedFilters.vue [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+      this.appliedFilters[this.appliedFilters.length - 1].query_operator = null;
+      this.$store.dispatch(
+        'contacts/setContactFilters',
+        JSON.parse(JSON.stringify(this.appliedFilters))
+      );
+      this.$emit('applyFilter', this.appliedFilters);
+    },
+    resetFilter(index, currentFilter) {
+      this.appliedFilters[index].filter_operator = this.filterTypes.find(
+        filter => filter.attributeKey === currentFilter.attribute_key
```
Comment (nithindavid) [inline app/javascript/dashboard/routes/dashboard/contacts/components/ContactsAdvancedFilters.vue:169]: One object has camel key and other has snake key. We could make them consistent

**Review by nithindavid — CHANGES_REQUESTED** [review]

File: app/javascript/dashboard/routes/dashboard/contacts/components/ContactsAdvancedFilters.vue (hunk 4 above)
Comment (fayazara) [reply app/javascript/dashboard/routes/dashboard/contacts/components/ContactsAdvancedFilters.vue:144]: The snake case convention is being used in the `/api/v1/accounts/1/contacts/filter`, the above array is the `v-model` data which I am sending as payload, hence the _

File: app/javascript/dashboard/routes/dashboard/contacts/components/Header.vue (hunk 5 above)
Comment (fayazara) [reply app/javascript/dashboard/routes/dashboard/contacts/components/Header.vue:187]: Fixed in https://github.com/chatwoot/chatwoot/pull/3471/commits/ec63d94adad3ec1d868806c97611996d7bba4e9a

File: app/javascript/dashboard/routes/dashboard/contacts/components/ContactsAdvancedFilters.vue (hunk 6 above)
Comment (fayazara) [reply app/javascript/dashboard/routes/dashboard/contacts/components/ContactsAdvancedFilters.vue:169]: This snake case convention being is used in the `/api/v1/accounts/1/contacts/filter` api, this key is being used as an identifier to reset the filter operator value, which is linked to the type of filter selected above.

Same goes with the `/conversation/filter` endpoint

File: app/javascript/dashboard/routes/dashboard/contacts/components/ContactsAdvancedFilters.vue (hunk 3 above)
Comment (fayazara) [reply app/javascript/dashboard/routes/dashboard/contacts/components/ContactsAdvancedFilters.vue]: Fixed in https://github.com/chatwoot/chatwoot/pull/3471/commits/3961c436c25de0f4c11c4a054d2d9d9931cc228c

**Review by nithindavid — DISMISSED** [review]
Code LGTM 👍

**Review by muhsin-k — DISMISSED** [review]

**Pull request merged**